Fix load more than 100 property groups and friends

JoshuaBehrens opened the pull request:
### 0. Sequel (not SQL)

This is a follow up to https://github.com/shopware/platform/pull/1508 of a follow up from https://github.com/shopware/platform/pull/1310 so @htkassner can pick it up more easily .

It differs in order of arguments so we always have criteria followed by an optional context and it is a rebase.

It now overrides iterateIds by @OliverSkroblin in the admin js repository class is overriden to contain my implementation from the previous PR.

It still does not fix self-paginating components like `sw-cms-el-config-product-listing`.

The changelog only mentions on issue as I don't know all of them.

### 1. Why is this change necessary?

There are lots of occassions where a complete list of objects is required but not requested properly. Most of these places are visible when you see a criteria limit of 100 or 500. These are tries to load all the data but this is not all the data. This might be the case because although it is tried to have php and admin-js behave identically regarding DAL operations it is not as the criteria in admin-js works with limited pages that have a preset of 1, 25 and the php criteria works with limit offset and has a preset of null, null. Those who work with the admin-js criteria and are familiar with it from php get fooled.

### 2. What does this change do, exactly?

Add the php-known RepositoryIterator and some constants from the criteria. Use the repository iterator at a lot of places that behave better now.
### 3. Please link to the relevant issues (if any).

https://issues.shopware.com/issues/NEXT-10262
https://issues.shopware.com/issues/NEXT-10349

And probably some more

### 4. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

JoshuaBehrens committed:
Add docs about RepositoryIterator in administration

JoshuaBehrens committed:
Use repository iterator to improve count requests

JoshuaBehrens committed:
Add changelog

mitelg commented on src/Administration/Resources/app/administration/src/module/sw-newsletter-recipient/page/sw-newsletter-recipient-list/index.js:
```diff
@@ -59,11 +59,11 @@ Component.register('sw-newsletter-recipient-list', {
             this.tagCollection = new EntityCollection('/tag', 'tag', Shopware.Context.api, new Criteria());
 
             const criteria = new Criteria(1, 100);
```
just a JS-noob question :grin: 

wouldn't it now makes sense to remove such occurrences of setting the limit to 100/500? they are just workarrounds, aren't they? 

or is there a benefit to leave them?

JoshuaBehrens replied:
It really is just to set the page size so we make bigger steps. When we start to refactor these calls for the page size we might also refactor all of them to keep https://github.com/shopware/platform/pull/2070 in mind. Not sure whether we should pull on all fuzzy threads :)

mitelg replied:
okay :+1: thanks for clearing this up

JoshuaBehrens commented:
New issue? I reference one in my branch name and in my changelog :D

jleifeld commented:
Hey @JoshuaBehrens ,

thanks for your MR. I checked it out but there are several things not working correctly:

- many Jest tests are failing
- many E2E tests are failing (reason for this is that many parts of the admin doesn't work correctly anymore)

Here are some screenshots where you can see some parts:

- Product listing does not work anymore:
![Bildschirmfoto_2021-10-04_um_08 23 39](https://user-images.githubusercontent.com/14929254/135804135-36a92ce5-718f-410f-8e88-27123402dcdf.png)

- Sales Channels aren't visible anymore:
![Bildschirmfoto_2021-10-04_um_08 23 45](https://user-images.githubusercontent.com/14929254/135804242-01fb5f45-7180-4985-9a3c-763993b39707.png)

- Category module:
![Bildschirmfoto_2021-10-04_um_08 23 51](https://user-images.githubusercontent.com/14929254/135804323-db01d2a2-b98e-4c71-ac76-6b47e203e94f.png)

- Options in property listing:
![Bildschirmfoto_2021-10-04_um_08 24 20](https://user-images.githubusercontent.com/14929254/135804385-04c8f515-2d23-452e-90f8-b9aa56ba20d3.png)

and more places.

Could you have a look why this happens?

Then I have also one question about this MR: what was the reason that you implemented the `iterate` method also in some listings with paginations?

JoshuaBehrens committed:
Add iterate and iterateId methods to repository iterators for easier usage

JoshuaBehrens committed:
Add repository iterator to admin js

JoshuaBehrens commented:
I rebased and fixed all the blunt technical issues I coudl find. Some things are still different to what one could expect. I touched listings as the property-whitelist configuration was bugging for one of our clients as they had too many properties and therefore the pagination was broken. I can look at how it will behave when I revert that change but this needs an other coding session. You can still have a look at the current behaviour. Maybe there are some other parts that behave weird that I didn't expect to have touched.

jleifeld commented:
Thanks for your effort! I will have a look at the MR when you are think it is working completely good

JoshuaBehrens commented:
Do you really had to migrate the criteria.data.js to criteria.data.ts now? 😖 this did make it a bit uneasy to rebase

JoshuaBehrens committed:
Use new total count mode constants from php in admin js

JoshuaBehrens committed:
Use repository iterator in places where lots of data is needed

Fixes NEXT-20262

JoshuaBehrens commented:
I found issues in how to use the new TotalCountMode constants so I included it into Shopware.Data. I also fixed the product listing. I hope I was able to find every issue but so far the admin works for me as I expect it.

htkassner commented:
@JoshuaBehrens As you already commented your initial problem is solved via https://github.com/shopware/platform/pull/2554. Due to this I will close this PR. Thank you for your effort! We combined this problem with a research to find places in the admin with similar problems to solve them all. We know there are several places with grids / listings without paging or search. We will address this! I will close this one here!
